meta-intel-iot-security/meta-security-smack/recipes-core/systemd/systemd/0001-core-smack-add-support-for-defining-smack-rules-path-v229.patch

```diff
@@ -0,0 +1,145 @@
+@@ -1607,6 +1622,7 @@ AC_MSG_RESULT([
+         Maximum System UID:      ${SYSTEM_UID_MAX}
+         Maximum System GID:      ${SYSTEM_GID_MAX}
+         Certificate root:        ${CERTIFICATEROOT}
++        Smack rules path:        ${DEFAULT_SMACK_RULES_PATH}
+ 
+         CFLAGS:                  ${OUR_CFLAGS} ${CFLAGS}
+         CPPFLAGS:                ${OUR_CPPFLAGS} ${CPPFLAGS}
+diff --git a/src/core/smack-setup.c b/src/core/smack-setup.c
+index 0c26e85..2515a30 100644
+--- a/src/core/smack-setup.c
++++ b/src/core/smack-setup.c
+@@ -38,6 +38,18 @@
+ 
+ #ifdef HAVE_SMACK
+ 
++#ifndef STRINGIFY
++#  define STRINGIFY(foo) #foo
++#endif
++
++#define DEFAULT_RULES_PATH_SMACK STRINGIFY(DEFAULT_SMACK_RULES_PATH) "/smack/accesses.d/"
++#define DEFAULT_RULES_PATH_CIPSO STRINGIFY(DEFAULT_SMACK_RULES_PATH) "/smack/cipso.d/"
++#define DEFAULT_RULES_PATH_NETLABEL STRINGIFY(DEFAULT_SMACK_RULES_PATH) "/smack/netlabel.d/"
++
++#define RULES_PATH_SMACK "/etc/smack/accesses.d/"
++#define RULES_PATH_CIPSO "/etc/smack/cipso.d/"
++#define RULES_PATH_NETLABEL "/etc/smack/netlabel.d/"
++
+ static int write_access2_rules(const char* srcdir) {
+         _cleanup_close_ int load2_fd = -1, change_fd = -1;
+         _cleanup_closedir_ DIR *dir = NULL;
+@@ -133,6 +145,7 @@ static int write_cipso2_rules(const char* srcdir) {
+         int r = 0;
+ 
+         cipso2_fd = open("/sys/fs/smackfs/cipso2", O_RDWR|O_CLOEXEC|O_NONBLOCK|O_NOCTTY);
++
+         if (cipso2_fd < 0)  {
+                 if (errno != ENOENT)
+                         log_warning_errno(errno, "Failed to open '/sys/fs/smackfs/cipso2': %m");
+@@ -274,16 +287,18 @@ int mac_smack_setup(bool *loaded_policy) {
+ 
+         assert(loaded_policy);
+ 
+-        r = write_access2_rules("/etc/smack/accesses.d/");
++        if ((r = write_access2_rules(RULES_PATH_SMACK)) == 0 ||
++            (r = write_access2_rules(DEFAULT_RULES_PATH_SMACK)) == 0)
++                log_info("Successfully loaded Smack policies.");
++
+         switch(r) {
+         case -ENOENT:
+                 log_debug("Smack is not enabled in the kernel.");
+                 return 0;
+         case ENOENT:
+-                log_debug("Smack access rules directory '/etc/smack/accesses.d/' not found");
++                log_debug("Smack access rules directory %s not found", DEFAULT_RULES_PATH_SMACK);
+                 return 0;
+         case 0:
+-                log_info("Successfully loaded Smack policies.");
+                 break;
+         default:
+                 log_warning_errno(r, "Failed to load Smack access rules, ignoring: %m");
+@@ -306,32 +321,36 @@ int mac_smack_setup(bool *loaded_policy) {
+                 log_warning_errno(r, "Failed to set SMACK netlabel rule \"127.0.0.1 -CIPSO\": %m");
+ #endif
+ 
+-        r = write_cipso2_rules("/etc/smack/cipso.d/");
++        if ((r = write_cipso2_rules(RULES_PATH_CIPSO)) == 0 ||
++            (r = write_cipso2_rules(DEFAULT_RULES_PATH_CIPSO)) == 0)
```

Comment:
Error handling could be better: if write_cipso2_rules(RULES_PATH_CIPSO) fails for any reason (not just "directory not found"), then the error is not logged. If r == -ENOENT, there's no point in trying to load from DEFAULT_RULES_PATH_SMACK.

IMHO we should only fall back to DEFAULT_RULES_PATH_SMACK for r == ENOENT.